Automated task grain size calculation (related to issue 567).

jhmkorndorfer opened the pull request:
This pull request includes the implementation for the automated task grain size selection in DAPHNE. Here we add a new option for partitioning strategy in DAPHNE called "AUTO". When a user selects AUTO, DAPHNE will automatically calculate a fixed task grain size which frequently achieves high performance.

@corepointer @pdamme This is my first pull request here. Please let me know if you need anything else from me.

corepointer commented:
Hi Jonas,

LGTM - looks good to me. If you want to write test cases for the scheduling code, parameterizing MultiThreadedKernelTest might be a good code sample to get inspired ;-) But I'll also merge it without the tests (after a manual test run).

Regards, Mark

corepointer reviewed (changes requested):
AUTO does not work for task sizes smaller than thread count. Please fix. I tried ```bin/daphne --vec --partitioning=AUTO --config UserConfig.json scripts/examples/hello-world.daph```

corepointer commented:
Since I rebased this branch on main and squashed the commits and had to force-push, you need to either move your local branch out of the way (e.g., rename it) and check out the remote one or reset your local branch to the remote version (losing any local modifications of that branch).

jhmkorndorfer commented:
> AUTO does not work for task sizes smaller than thread count. Please fix. I tried `bin/daphne --vec --partitioning=AUTO --config UserConfig.json scripts/examples/hello-world.daph`

Good catch! I will get to it ASAP after christmas.

jhmkorndorfer commented:
> Since I rebased this branch on main and squashed the commits and had to force-push, you need to either move your local branch out of the way (e.g., rename it) and check out the remote one or reset your local branch to the remote version (losing any local modifications of that branch).

Hm.. I am not sure I understand. What exactly should I do?

corepointer commented:
> Hm.. I am not sure I understand. What exactly should I do?

You will get weird errors if you try to pull the changes I made to this branch (because I had to force-push after rebase+squash) to your local copy. If you did not commit any changes to your local version of this branch, then just get rid of it, pull the remote version and work from there.

jhmkorndorfer commented:
Hello @corepointer I have added a fix to the issue regarding "AUTO does not work for task sizes smaller than thread count". I hope I did not mess up anything with the branch hehe. I have just reset it to the remote version and then pushed the fix.

corepointer commented:
LGTM - thx for fixing the little issue. I integrated it in the main commit and merged to main.
